src/main/java/io/confluent/connect/elasticsearch/DataConverter.java

```diff
@@ -152,51 +144,71 @@ public IndexableRecord convertRecord(
           );
           break;
         case FAIL:
-          throw new DataException(String.format(
-              "Sink record with key of %s and null value encountered for topic/partition/offset "
-              + "%s/%s/%s (to ignore future records like this change the configuration property "
-              + "'%s' from '%s' to '%s')",
-              record.key(),
-              record.topic(),
-              record.kafkaPartition(),
-              record.kafkaOffset(),
-              ElasticsearchSinkConnectorConfig.BEHAVIOR_ON_NULL_VALUES_CONFIG,
-              BehaviorOnNullValues.FAIL,
-              BehaviorOnNullValues.IGNORE
-          ));
+          throw new DataException(
+              String.format(
+                  "Sink record with key of %s and null value encountered for topic/partition/offset"
+                      + " %s/%s/%s (to ignore future records like this change the configuration"
+                      + " property '%s' from '%s' to '%s')",
+                  record.key(),
+                  record.topic(),
+                  record.kafkaPartition(),
+                  record.kafkaOffset(),
+                  ElasticsearchSinkConnectorConfig.BEHAVIOR_ON_NULL_VALUES_CONFIG,
+                  BehaviorOnNullValues.FAIL,
+                  BehaviorOnNullValues.IGNORE
+              )
+          );
         default:
-          throw new RuntimeException(String.format(
-              "Unknown value for %s enum: %s",
-              BehaviorOnNullValues.class.getSimpleName(),
-              behaviorOnNullValues
-          ));
+          throw new ConnectException(
+              String.format(
+                  "Unknown value for %s enum: %s",
+                  BehaviorOnNullValues.class.getSimpleName(),
+                  config.behaviorOnNullValues()
+              )
+          );
       }
     }
 
-    final String id;
-    if (ignoreKey) {
-      id = record.topic()
-           + "+" + String.valueOf((int) record.kafkaPartition())
-           + "+" + String.valueOf(record.kafkaOffset());
-    } else {
-      id = convertKey(record.keySchema(), record.key());
+    final String payload = getPayload(record);
+    final String id = config.shouldIgnoreKey(record.topic())
+        ? String.format("%s+%d+%d", record.topic(), record.kafkaPartition(), record.kafkaOffset())
+        : convertKey(record.keySchema(), record.key());
+
+    if (record.value() != null) {
+      if (config.writeMethod() == WriteMethod.UPSERT) {
+        return new UpdateRequest(index, id)
+            .doc(payload, XContentType.JSON)
+            .upsert(payload, XContentType.JSON)
+            .retryOnConflict(Math.max(config.maxInFlightRequests(), 5));
+      }
+
+      return addExternalVersioning(
```

Comment:
Maybe add a comment or use a `switch` to handle the `WriteMethod` enum so that it's more obvious that we are handling `INSERT`?